LSP Code Lens

mholo65 opened the pull request:
Adds CodeLensHandler and CodeLensResolveHandler

mholo65 commented on src/OmniSharp.LanguageServerProtocol/Handlers/CodeLensHandler.cs:
```diff
@@ -0,0 +1,128 @@
+            foreach (var (selector, membersAsTreeHandler, findUsagesHandler) in handlers
+                .OfType<
+                    Mef.IRequestHandler<MembersTreeRequest, FileMemberTree>,
+                    Mef.IRequestHandler<FindUsagesRequest, QuickFixResponse>>())
```
After adding this (in second commit) this method will now return null. Haven't had time to debug yet. Without the resolver functionalty, the code lens handler works ok. @david-driscoll do you spot any errors which could lead to this? The same handler is used in #1070

david-driscoll commented:
I'm not sure what could be causing that... what happens if you reverse the two? and do find usages first?

mholo65 commented:
@david-driscoll tried reversing. Still same issue.. I've debugged and see that I get an exception. 

It throws at `_server.AddHandlers(CodeLensHandler.Enumerate(_handlers));` with details:

```
System.ArgumentNullException: Value cannot be null.
   at System.Reflection.RuntimeMethodInfo.MakeGenericMethod(Type[] methodInstantiation)
   at OmniSharp.Extensions.LanguageServer.HandlerCollection.Add(IJsonRpcHandler[] handlers)
   at OmniSharp.Extensions.LanguageServer.LanguageServer.AddHandlers(IJsonRpcHandler[] handlers)
   at OmniSharp.Extensions.LanguageServer.LanguageServer.AddHandlers(IEnumerable`1 handlers)
   at OmniSharp.LanguageServerProtocol.LanguageServerHost.Initialize(InitializeParams initializeParams) in C:\Users\mb\src\gh\omnisharp-roslyn\src\OmniSharp.LanguageServerProtocol\LanguageServerHost.cs:line 162
```

Know to little about the C# LSP internals to know why it throws. Do you might know what could cause this?

mholo65 commented:
@david-driscoll is it because `ICodeLensResolveHandler` doesn't inherit `IRegistration<T>` and/or `ICapability<T>`...?

mholo65 commented:
Waiting for https://github.com/OmniSharp/csharp-language-server-protocol/pull/70

mholo65 commented:
@david-driscoll with https://github.com/OmniSharp/csharp-language-server-protocol/pull/81 I'll managed to get this working. Now I only need to figure out the correct commands to pass back to client...

Should we hard code for VSCode at first, or what should we do?

mholo65 commented on src/OmniSharp.LanguageServerProtocol/Handlers/CodeLensHandler.cs:
```diff
@@ -0,0 +1,140 @@
+            {
+                Title = length == 1 ? "1 reference" : $"{length} references",
+                Name = "editor.action.showReferences",
+                Arguments = new JArray
```
VSCode is currently not accepting this... Need to figure out the commands

mholo65 commented on src/OmniSharp.LanguageServerProtocol/Handlers/CodeLensHandler.cs:
```diff
@@ -0,0 +1,140 @@
+            var textDocumentUri = value.Data.ToObject<Uri>();
+
+            return textDocumentUri != null &&
+                _documentSelector.IsMatch(new TextDocumentAttributes(textDocumentUri, string.Empty));
```
Dunno if this is the best approach. However, this is not called since ResolveHandler is the same as the CommandHandler

mholo65 commented:
Rebased and fixed. Now works, no commands yet.

![image](https://user-images.githubusercontent.com/7863439/53200795-19736f80-362b-11e9-9769-f255202500c0.png)
